### PR TITLE
std: Don't pass overlapped handles to processes

### LINKED ISSUE #38811: std::process::Command: piped stderr hangs in Windows
It looks like the use of overlapped I/O on stdio ports may cause some problems in Windows. When I try to shell out to `npm install` using `Stdio::piped()` on the stderr port, the command runs successfully but hangs at the end.

@alexcrichton pointed me to this commit as the likely culprit: https://github.com/rust-lang/rust/commit/7c3038f82477491e20c6f80c0139ddb1f1b912ca

Here's the smallest test case I could figure out how to create: https://github.com/dherman/hang-command

I wanted to try to come up with a minimal Node script to shell out to, as opposed to `npm install`. Unfortunately, I'm not good enough at debugging npm and Node to figure out exactly what they're doing that causes the hang.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit fixes a mistake introduced in #31618 where overlapped handles were
leaked to child processes on Windows. On Windows once a handle is in overlapped
mode it should always have I/O executed with an instance of `OVERLAPPED`. Most
child processes, however, are not prepared to have their stdio handles in
overlapped mode as they don't use `OVERLAPPED` on reads/writes to the handle.

Now we haven't had any odd behavior in Rust up to this point, and the original
bug was introduced almost a year ago. I believe this is because it turns out
that if you *don't* pass an `OVERLAPPED` then the system will [supply one for
you][link]. In this case everything will go awry if you concurrently operate on
the handle. In Rust, however, the stdio handles are always locked, and there's
no way to not use them unlocked in libstd. Due to that change we've always had
synchronized access to these handles, which means that Rust programs typically
"just work".

Conversely, though, this commit fixes the test case included, which exhibits
behavior that other programs Rust spawns may attempt to execute. Namely, the
stdio handles may be concurrently used and having them in overlapped mode wreaks
havoc.

[link]: https://blogs.msdn.microsoft.com/oldnewthing/20121012-00/?p=6343

Closes #38811

### CODE REVIEW

File: src/libstd/sys/windows/pipe.rs [hunk 1]
```diff
@@ -54,11 +79,16 @@ pub fn anon_pipe() -> io::Result<(AnonPipe, AnonPipe)> {
                                   .encode_wide()
                                   .chain(Some(0))
                                   .collect::<Vec<_>>();
+            let mut flags = c::FILE_FLAG_FIRST_PIPE_INSTANCE |
+                c::FILE_FLAG_OVERLAPPED;
```
Comment (nagisa) [inline src/libstd/sys/windows/pipe.rs]: Indentation is weird here. I immediately assumed this was inside some conditional.

Comment (nagisa) [inline src/libstd/sys/windows/pipe.rs]: I’d honestly prefer if it was written as such:

```
let access_flag = if ours_readable { PIPE_ACCESS_INBOUND } else { PIPE_ACCESS_OUTBOUND };
let handle = c::CreateNamedPipeW(wide_name.as_ptr(),
                                                         PIPE_1 | 
                                                         PIPE_2 | 
                                                         access_flags, ...
```

**Comment by alexcrichton** [comment]
@bors: p=1

This is a backport

**Comment by arielb1** [comment]
```
The command "if [ "$ALLOW_PR" = "" ] && [ "$TRAVIS_BRANCH" != "auto" ]; then

    echo skipping, not a full build;

elif [ "$TRAVIS_OS_NAME" = "osx" ]; then

    git submodule update --init &&

    src/ci/run.sh;

else

    git submodule update --init &&

    src/ci/docker/run.sh $IMAGE;

fi

" exited with 129.
```

@bors retry

**Comment by alexcrichton** [comment]
@bors: retry

* osx linker segfaulted

**Pull request merged**